pkg/sbom/package.go

```diff
@@ -0,0 +1,191 @@
+// into a private, generalized bill of materials model (with relationship data)
+// designed to be converted to specific formats — for now, just SPDX.
+package sbom
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"chainguard.dev/apko/pkg/sbom/generator/spdx"
+	"github.com/chainguard-dev/clog"
+	"github.com/github/go-spdx/v2/spdxexp"
+	purl "github.com/package-url/packageurl-go"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+)
+
+// Package is a representation of an SBOM package specified by the build
+// process. It is later converted to an SPDX package, but it doesn't expose
+// fields that are invariant in the SPDX output.
+type Package struct {
+	// The name of the origin package, a subpackage, or any other kind of (e.g.
+	// non-APK) package for inclusion in the SBOM.
+	Name string
+
+	// The version of the package. For APK packages, this should be the "full
+	// version" (including the epoch).
+	Version string
+
+	// This is the copyright text in the SPDX package. It's usually left blank.
+	Copyright string
+
+	// SPDX license expression. Leaving this empty will result in NOASSERTION being
+	// used as its value.
+	LicenseDeclared string
+
+	// Name of the distro/organization that produced the package. E.g. "wolfi".
+	//
+	// TODO: consider renaming this to avoid confusion from our other uses of
+	//  "namespace", perhaps to "supplier" or "originator" (or have both), and signal
+	//  that it's safe to leave this blank.
+	Namespace string
+
+	// The architecture of the package. E.g. "aarch64". This field isn't always
+	// relevant, especially when describing material upstream of the built APK
+	// package (e.g. source code or language ecosystem dependencies).
+	Arch string
+
+	// Checksums of the package. The keys are the checksum algorithms (e.g. "SHA-256"),
+	// and the values are the checksums.
+	//
+	// TODO: We're not currently using this field, consider removing it.
+	Checksums map[string]string
+
+	// The Package URL for this package, if any. If set, it will be added as the
+	// only ExternalRef of type "purl" to the SPDX package. (A package
+	// should have only one PURL external ref.)
+	PURL *purl.PackageURL
+}
+
+// ToSPDX returns the Package converted to its SPDX representation.
+func (p Package) ToSPDX(ctx context.Context) spdx.Package {
+	log := clog.FromContext(ctx)
+
+	if p.LicenseDeclared == "" {
+		log.Warnf("%s: no license specified, defaulting to %s", p.ID(), spdx.NOASSERTION)
+		p.LicenseDeclared = spdx.NOASSERTION
+	} else {
+		valid, bad := spdxexp.ValidateLicenses([]string{p.LicenseDeclared})
+		if !valid {
+			log.Warnf("invalid license: %s", strings.Join(bad, ", "))
+		}
+	}
+
+	sp := spdx.Package{
+		ID:               p.ID(),
+		Name:             p.Name,
+		Version:          p.Version,
+		FilesAnalyzed:    false,
+		LicenseConcluded: spdx.NOASSERTION,
+		LicenseDeclared:  p.LicenseDeclared,
+		DownloadLocation: spdx.NOASSERTION,
+		CopyrightText:    p.Copyright,
```

Comment:
i think currently somehow copyright ends up "\n" which confused one customer. if it is optional field can we omit it? as we have no way to specify it or to scan it / populate it.

Reply:
Yeah agreed

Reply:
also not a regression if it stays the same.